### PR TITLE
Do codegen and check no-diff of all (non-blacklisted) CUE->TS codegen during CI

### PR DESCRIPTION
**Pull request by sdboyer**
This PR introduces a new build step that automatically runs the CUE->TS code generator - `grafana-cli cue gen-ts` - and ensures that there's no diffs/untracked files afterwards. The effect is a CI check that ensures the generated TS is always in sync with the all input CUE files.

We also introduce a blacklist to exclude files from codegen, to make it easier to tackle this codegen one bundle of CUE files at a time. The blacklist removes CUE files from what is considered to be "all input CUE files" above. The contents of the blacklist include all CUE files that we would expect to translate to TS.

In other words, this PR is kind of a no-op :)

We're doing this to create a ratchet, and lay the foundation for future PRs. The intention is that future PRs will be able to remove ~one line at a time from the blacklist, and focus the discussion there on the resulting TS file that starts getting generated by `grafana-cli cue gen-ts`. Then, once such a PR is merged, the CI test that's already in place "ratchets" up, ensuring on every subsequent CI run that the now-unblacklisted CUE->TS translation is exactly the same as what's committed.

### CODE REVIEW

**Comment by sdboyer** [comment]
@dsotirakis seems like i've done something unintentionally horrifying to the `.drone.yml` by running `make drone` - help?

**Comment by dsotirakis** [comment]
@sdboyer I think that's because you have the latest `drone-cli` version installed.
(running `drone -v` locally on your machine will result in `1.4.0` I think) - try and downgrade to `1.2.4`!

**Comment by sdboyer** [comment]
ah, i have 1.3.1, will downgrade, ty!

File: pkg/cmd/grafana-cli/commands/cuetsify_command.go [hunk 1]
```diff
@@ -27,12 +27,39 @@ import (
...
+// Hard-coded list of paths to skip. Uncomment a particular file as we're ready
+// to rely on the TypeScript auto-generated by cuetsy for that particular file.
+var skipPaths = []string{
+	"public/app/plugins/panel/barchart/models.cue",
+	"public/app/plugins/panel/bargauge/models.cue",
+	"public/app/plugins/panel/canvas/models.cue",
+	"public/app/plugins/panel/dashlist/models.cue",
+	"public/app/plugins/panel/gauge/models.cue",
+	"public/app/plugins/panel/histogram/models.cue",
+	"public/app/plugins/panel/news/models.cue",
```
Comment (ryantxu) [inline pkg/cmd/grafana-cli/commands/cuetsify_command.go:46]: ^^ news should be good to run also!

File: .drone.yml [hunk 2]
```diff
@@ -107,6 +108,19 @@ steps:
...
   - build-backend
 
+- name: ensure-cuetsified
+  image: grafana/build-container:1.4.3
+  commands:
+  - ./bin/linux-amd64/grafana-cli cue gen-ts --grafana-root .
+  - "# The above command generates Typescript files (*.gen.ts) from all appropriate .cue files."
+  - "# It is required that the generated Typescript be in sync with the input CUE files."
+  - "# ...Modulo eslint auto-fixes...:"
+  - ./node_modules/.bin/eslint . --ext .gen.ts --fix
```
Comment (sdboyer) [inline .drone.yml:118]: Is this an acceptable way to invoke our linter? Seems to be what we do with husky in `package.json`

File: scripts/drone/steps/lib.star [hunk 3]
```diff
@@ -1023,3 +1024,21 @@ def validate_scuemata_step():
...
+        'image': build_image,
+        'depends_on': [
+            'validate-scuemata',
+        ],
+        'commands': [
+            './bin/linux-amd64/grafana-cli cue gen-ts --grafana-root .',
+            '# The above command generates Typescript files (*.gen.ts) from all appropriate .cue files.',
+            '# It is required that the generated Typescript be in sync with the input CUE files.',
+            '# ...Modulo eslint auto-fixes...:',
+            './node_modules/.bin/eslint . --ext .gen.ts --fix',
```
Comment (sdboyer) [inline scripts/drone/steps/lib.star:1040]: Is this an acceptable way to invoke typescript linting? Seems to be what we do with husky in package.json

File: scripts/drone/steps/lib.star [hunk 4]
```diff
@@ -487,6 +487,7 @@ def codespell_step():
             # Important: all words have to be in lowercase, and separated by "\n".
             'echo -e "unknwon\nreferer\nerrorstring\neror\niam\nwan" > words_to_ignore.txt',
             'codespell -I words_to_ignore.txt docs/',
+            'rm words_to_ignore.txt',
```
Comment (sdboyer) [inline scripts/drone/steps/lib.star:490]: must remove, otherwise it hangs over to the other pipeline and the git cleanness checker fails

**Comment by sdboyer** [comment]
This should now be ready. I've taken `news` off the blacklist in this PR, but otherwise, all existing panel plugins are blacklisted for easy iterative processing in future PRs.

@grafana/user-essentials @ryantxu PTAL

File: scripts/drone/steps/lib.star (hunk 3 above)
Comment (hugohaggmark) [reply scripts/drone/steps/lib.star:1040]: Not sure if prettier (gofmt for Typescript) is run by cuetsy, otherwise "prettier --write" might be a good idea to run to keep all formatting the same?

File: scripts/drone/steps/lib.star (hunk 4 above)
Comment (hugohaggmark) [reply scripts/drone/steps/lib.star:490]: Makes sense 👍

File: scripts/drone/steps/lib.star (hunk 3 above)
Comment (sdboyer) [reply scripts/drone/steps/lib.star:1040]: Yeah this is what confused me - @ryantxu mentioned using yarn [here](https://github.com/grafana/grafana/pull/39694#issuecomment-928725349), but we use `eslint` directly in pre-commit, so that strikes me as the truest arbiter of what the expected final committed state.

**Comment by sdboyer** [comment]
Seems like the particular way we invoke TS linting doesn't matter so much, not enough to block progress.

Since the real intent here is to introduce the foundations of a workflow for @grafana/user-essentials, i'm gonna hope one of you all are willing to do the official PR approval 😄

File: scripts/drone/steps/lib.star (hunk 3 above)
Comment (hugohaggmark) [reply scripts/drone/steps/lib.star:1040]: `so that strikes me as the truest arbiter of what the expected final committed state` as English is not my first language, the wordings in this sentence makes me translate the individual words so I can understand the underlying meaning. Not sure I understand tbh, can you use simpler words?

**Comment by hugohaggmark** [comment]
> Seems like the particular way we invoke TS linting doesn't matter so much, not enough to block progress.
> 
> Since the real intent here is to introduce the foundations of a workflow for @grafana/user-essentials, i'm gonna hope one of you all are willing to do the official PR approval 😄

@grafana/user-essentials Sam is asking for our feedback here

File: scripts/drone/steps/lib.star (hunk 3 above)
Comment (ashharrison90) [reply scripts/drone/steps/lib.star:1040]: so looking at package.json it looks like we only use `prettier` on `{json,scss,md,mdx}` files. that surprises me, i thought we ran it for everything. in that case, i think what you've got is fine for now (it's the same as what we do to all the other typescript files anyway)

File: pkg/cmd/grafana-cli/commands/cuetsify_command.go [hunk 5]
```diff
@@ -27,12 +27,38 @@ import (
...
 var ctx = cuecontext.New()
 
+// The only import statement we currently allow in any models.cue file
 const allowedImport = "github.com/grafana/grafana/packages/grafana-schema/src/schema"
 
 var importMap = map[string]string{
 	allowedImport: "@grafana/schema",
 }
 
+// Hard-coded list of paths to skip. Uncomment a particular file as we're ready
```
Comment (ashharrison90) [inline pkg/cmd/grafana-cli/commands/cuetsify_command.go:37]: `Remove a particular file...` or `Comment out a particular file...`?
```suggestion
// Hard-coded list of paths to skip. Remove a particular file as we're ready
```

**Review by ashharrison90 — APPROVED** [review]
minor comment change, otherwise lgtm

File: scripts/drone/steps/lib.star [hunk 6]
```diff
@@ -1023,3 +1024,21 @@ def validate_scuemata_step():
...
+        'depends_on': [
+            'validate-scuemata',
+        ],
+        'commands': [
+            './bin/linux-amd64/grafana-cli cue gen-ts --grafana-root .',
+            '# The above command generates Typescript files (*.gen.ts) from all appropriate .cue files.',
+            '# It is required that the generated Typescript be in sync with the input CUE files.',
+            '# ...Modulo eslint auto-fixes...:',
+            './node_modules/.bin/eslint . --ext .gen.ts --fix',
+            '# If any filenames are emitted by the below script, run the generator command locally and commit the result.',
```
Comment (axelavargas) [inline scripts/drone/steps/lib.star:1041]: 🤔 It can be repeating ourselves here, but I think we could add the actual command on the comment, then is easier to just copy and paste it in the terminal
```suggestion
            '# If any filenames are emitted by the below script, run the generator command `grafana-cli cue gen-ts` locally and commit the result.',
```

**Review by axelavargas — APPROVED** [review]
Hallo, @sdboyer nice job on this PR 👏🏾. Thank you for adding detailed comments in the scripts, it was easier to follow the logic 🙌🏾 , I left a minor comment, but besides that it looks good to me.

File: scripts/drone/steps/lib.star (hunk 6 above)
Comment (sdboyer) [reply scripts/drone/steps/lib.star:1041]: sure!

**Review by ashharrison90 — CHANGES_REQUESTED** [review]
hey @sdboyer, quick question.

user essentials currently doesn't have the bandwidth to look at doing the generation stuff, at least not any time soon. what happens if tomorrow the CI starts failing on this particular step? is the intention that user essentials would own this code/need to fix that, or is that something you would be doing?

**Review by ashharrison90 — APPROVED** [review]
lgtm 🙌
